src/hl_memory.c

```diff
@@ -6,101 +6,171 @@
-	if (status != synSuccess) {
-		fprintf(stderr, "Failed to get device info, status %d\n", status);
+	memset(hl_ctx->mem_handle_table, 0, sizeof(hl_ctx->mem_handle_table));
+
+	if (pthread_mutex_init(&hl_ctx->mem_handle_table_lock, NULL) != 0) {
+		(void) hlthunk_close(hl_ctx->device_fd);
 		return FAILURE;
 	}
 
-	hl_ctx->device_fd = device_info.fd;
 	return SUCCESS;
 }
 
-int hl_memory_destroy(struct memory_ctx *ctx) {
-	struct hl_memory_ctx *hl_ctx = container_of(ctx, struct hl_memory_ctx, base);
+int hl_memory_destroy(struct memory_ctx *ctx)
+{
+	struct hl_memory_ctx *const hl_ctx = container_of(ctx, struct hl_memory_ctx, base);
 
-	synDeviceRelease(hl_ctx->device_id);
-	synDestroy();
+	(void) pthread_mutex_destroy(&hl_ctx->mem_handle_table_lock);
+	(void) hlthunk_close(hl_ctx->device_fd);
 
 	free(hl_ctx);
 	return SUCCESS;
 }
 
 int hl_memory_allocate_buffer(struct memory_ctx *ctx, int alignment, uint64_t size, int *dmabuf_fd,
-			      uint64_t *dmabuf_offset, void **addr, bool *can_init) {
-	struct hl_memory_ctx *hl_ctx = container_of(ctx, struct hl_memory_ctx, base);
-	int fd;
-	uint64_t buffer_addr;
-	size_t buf_size = (size + ACCEL_PAGE_SIZE - 1) & ~(ACCEL_PAGE_SIZE - 1);
-	synStatus status = synDeviceMalloc(hl_ctx->device_id, buf_size, 0, 0, &buffer_addr);
-
-	if (status != synSuccess) {
-		fprintf(stderr, "Failed to allocate HL memory on device %d of size %lu\n",
-			hl_ctx->device_id, (unsigned long)buf_size);
+							  uint64_t *dmabuf_offset, void **addr, bool *can_init)
+{
+	struct hl_memory_ctx *const hl_ctx = container_of(ctx, struct hl_memory_ctx, base);
+	const uint64_t page_size = 0;
+	const uint64_t NO_OFFSET = 0;
+	const bool NOT_SHARED = false;
+	int fd = INVALID_FD;
+	uint64_t buffer_addr = 0;
+	const size_t buf_size = (size + ACCEL_PAGE_SIZE - 1) & ~(ACCEL_PAGE_SIZE - 1);
+
+	const uint64_t memory_handle = hlthunk_device_memory_alloc(hl_ctx->device_fd, buf_size, page_size,
+															   HL_MEM_CONTIGUOUS, NOT_SHARED);
+	if (memory_handle == 0) {
+		fprintf(stderr, "Failed to allocate %lu bytes of device memory\n", buf_size);
+		return FAILURE;
+	}
+	buffer_addr = hlthunk_device_memory_map(hl_ctx->device_fd, memory_handle, 0);
+	if (buffer_addr == 0) {
+		fprintf(stderr, "Failed to map device memory allocation\n");
+		return FAILURE;
+	}
+	if (pthread_mutex_lock(&hl_ctx->mem_handle_table_lock) != 0) {
+		fprintf(stderr, "Failed to lock mutex while allocating memory\n");
+		return FAILURE;
+	}
+	if (hl_set_memory_handle(hl_ctx, buffer_addr, memory_handle) != SUCCESS) {
+		(void) pthread_mutex_unlock(&hl_ctx->mem_handle_table_lock);
+		return FAILURE;
+	}
+	if (pthread_mutex_unlock(&hl_ctx->mem_handle_table_lock) != 0) {
+		fprintf(stderr, "Failed to unlock mutex\n");
 		return FAILURE;
 	}
 
-	fd = hlthunk_device_memory_export_dmabuf_fd(hl_ctx->device_fd, buffer_addr, buf_size, 0);
+	fd = hlthunk_device_mapped_memory_export_dmabuf_fd(hl_ctx->device_fd, buffer_addr, buf_size, NO_OFFSET,
```

Comment:
is this code supposed to be backward compatible to gaudi 1?
if so, this call was changed, need to call `hlthunk_device_memory_export_dmabuf_fd` if the device is gaudi 1

(very confusing names, hlthunk_device_memory_export_dmabuf_fd vs hlthunk_device_mapped_memory_export_dmabuf_fd)